File: backend/app/models/signal.py
from pydantic import BaseModel, UUID4
from datetime import datetime
from typing import Literal, Optional


class SignalModel(BaseModel):
    id: UUID4
    scan_id: UUID4
    symbol: str
    category: str
    # User-facing labels (populated by router layer, not stored in DB)
    display_status: Optional[Literal["BUY", "BREAKOUT", "WATCH", "NO_ACTION"]] = None
    display_category: Optional[str] = None  # "Long-Term Buy", "Swing Buy", etc.
    side: Optional[str] = None
    signal_timeframe: Optional[str] = None
    sl_timeframe: Optional[str] = None
    trend_direction: Optional[str] = None
    entry: Optional[float] = None
    stop_loss: Optional[float] = None
    sl_distance_pct: Optional[float] = None
    t1: Optional[float] = None
    t2: Optional[float] = None
    t3: Optional[float] = None
    rr_t1: Optional[float] = None
    rr_t2: Optional[float] = None
    rr_t3: Optional[float] = None
    gate_strength: Optional[float] = None
    volatility_compression: Optional[float] = None
    breakout_probability: Optional[float] = None
    confidence: Optional[float] = None
    rank_score: Optional[float] = None
    mtf_alignment_pct: Optional[float] = None
    structure_quality: Optional[float] = None
    atr: Optional[float] = None
    htf_confirmed: Optional[bool] = None
    correction_validated: Optional[bool] = None
    bounce_sequence_valid: Optional[bool] = None
    fib_confluence: Optional[bool] = None
    phase: Optional[str] = None
    trailing_plan: Optional[dict] = None
    reasoning: Optional[str] = None
    # ---- strategy-rework fields (migration 008) ----
    breakout_state: Optional[str] = None
    range_high: Optional[float] = None
    range_low: Optional[float] = None
    breakout_level: Optional[float] = None
    measured_move: Optional[float] = None
    rs_score: Optional[float] = None
    sector_momentum: Optional[float] = None
    accumulation_score: Optional[float] = None
    fundamental_score: Optional[float] = None
    volume_buildup: Optional[bool] = None
    created_at: datetime


class SignalFilters(BaseModel):
    category: Optional[str] = None
    min_rank: float = 0
    min_gate: float = 0
    side: Optional[str] = None
    timeframe: Optional[str] = None
    limit: int = 50
    offset: int = 0


class SignalListResponse(BaseModel):
    total: int
    items: list[SignalModel]

File: apps/web/src/types/signal.ts
export type SignalCategory = "INVESTMENT" | "SWING" | "POSITIONAL" | "BREAKOUT" | "WATCH" | "IGNORE";
export type SignalSide = "BUY" | "SELL";
export type DisplayStatus = "BUY" | "BREAKOUT" | "WATCH" | "NO_ACTION";
export type BreakoutState =
  | "BUY_ZONE"
  | "BREAKOUT_CONFIRMED"
  | "ACCUMULATION"
  | "EXTENDED"
  | "BROKEN_DOWN"
  | "NO_GATE";

export interface Signal {
  id: string;
  scan_id: string;
  symbol: string;
  category: SignalCategory;
  // User-facing labels added by the API layer
  display_status: DisplayStatus | null;
  display_category: string | null;  // "Long-Term Buy", "Swing Buy", "Positional Buy", "Watch", "No Action"
  side: SignalSide | null;
  signal_timeframe: string | null;
  sl_timeframe: string | null;
  trend_direction: string | null;
  entry: number | null;
  stop_loss: number | null;
  sl_distance_pct: number | null;
  t1: number | null;
  t2: number | null;
  t3: number | null;
  rr_t1: number | null;
  rr_t2: number | null;
  rr_t3: number | null;
  gate_strength: number | null;
  volatility_compression: number | null;
  breakout_probability: number | null;
  confidence: number | null;
  rank_score: number | null;
  mtf_alignment_pct: number | null;
  structure_quality: number | null;
  atr: number | null;
  htf_confirmed: boolean | null;
  correction_validated: boolean | null;
  bounce_sequence_valid: boolean | null;
  fib_confluence: boolean | null;
  phase: string | null;
  trailing_plan: {
    on_T1_hit?: string;
    on_T2_hit?: string;
    on_T3_hit?: string;
  } | null;
  reasoning: string | null;
  // Strategy-rework fields (consolidation-range / breakout-state model)
  breakout_state: BreakoutState | null;
  range_high: number | null;
  range_low: number | null;
  breakout_level: number | null;
  measured_move: number | null;
  rs_score: number | null;
  sector_momentum: number | null;
  accumulation_score: number | null;
  fundamental_score: number | null;
  volume_buildup: boolean | null;
  created_at: string;
  // Enriched from stock_master JOIN (present when stock_master is populated)
  company_name?: string | null;
  sector?: string | null;
}

export interface SignalListResponse {
  total: number;
  items: Signal[];
}

export interface SignalFilters {
  category?: SignalCategory;
  min_rank?: number;
  min_gate?: number;
  side?: SignalSide;
  // timeframe is always "1d" in this platform — not a user-facing filter
  limit?: number;
  offset?: number;
}
